pt_miniscreen/core/component.py

```diff
@@ -0,0 +1,223 @@
+
+    @property
+    def input(self):
+        return self._input.copy()
+
+    @input.setter
+    def input(self, next_input):
+        self._input = next_input.copy()
+
+    @property
+    def output(self):
+        return self._output.copy()
+
+    @output.setter
+    def output(self, next_output):
+        self._output = next_output.copy()
+
+
+class Component:
+    default_state: Dict[Any, Any] = {}
+
+    def __del__(self):
+        logger.debug(f"Garbage collect {self}")
+        self._cleanup()
+
+    def __init__(self, on_rerender, initial_state={}):
+        self._children = []
+        self._intervals = []
+        self._render_cache = RenderCache()
+        self._get_on_rerender = WeakMethod(on_rerender)
+        self.state = State(
+            initial_state={**self.default_state, **initial_state},
+            on_state_update=self._on_state_update,
+        )
+        self._reconciliation_lock = threading.Lock()
+        self._reconciliation_queued = False
+
+        # wrap subclass render method to track image properties and optimise
+        self._unmodified_render = self.render
+
+        def render(image):
+            # set size of component to input image for use in calculations
+            self.size = image.size
+            self.width = image.width
+            self.height = image.height
+
+            # return cached output if input is the same
+            if is_same_image(image, self._render_cache.input):
+                return self._render_cache.output
+
+            logger.debug(f"{self} rendering")
+            self._render_cache.input = image
+            output = self._unmodified_render(image.copy())
+            self._render_cache.output = output
+            return output
+
+        self.render = render
+
+    def _cleanup(self):
+        # stop notifying parent of updates immediately
+        self._get_on_rerender = lambda: None
+
+        # subclass specific cleanup
+        self.cleanup()
+
+        for interval in self._intervals:
+            interval.cancel()
+
+        for child in self._children:
+            child._cleanup()
+
+        self._intervals = []
+        self._children = []
+
+    def _reconcile(self):
+        if self._reconciliation_queued:
+            # since state is always up-to-date we can let the queued reconcile
+            # handle the rerender and ignore this one. This prevents the queue
+            # building up with very rapid updates
+            return
+
+        try:
+            self._reconciliation_queued = self._reconciliation_lock.locked()
+            self._reconciliation_lock.acquire()
+
+            # bail if parent no longer exists
+            on_rerender = self._get_on_rerender()
+            if not callable(on_rerender):
+                return
+
+            # if render has never been called always notify parent about rerender
+            if not isinstance(self._render_cache.input, Image.Image):
+                return on_rerender()
```

Comment:
I've added a comment explaining why it's necessary to call `on_rerender`. I've oscillated between whether to do it or not but since this is intended to be used by people with a wide variety of coding experience I settled on the more forgiving option.